packages/dev/core/src/Materials/Textures/Procedurals/proceduralTexture.ts

```diff
@@ -703,25 +711,29 @@ export class ProceduralTexture extends Texture {
             for (let layer = 0; layer < numLayers; layer++) {
                 engine.bindFramebuffer(this._rtWrapper, 0, undefined, undefined, true, 0, layer);
 
+                // VBOs
+                engine.bindBuffers(this._vertexBuffers, this._indexBuffer, this._drawWrapper.effect!);
+
                 if (this._rtWrapper.is3D || this._rtWrapper.is2DArray) {
                     this._drawWrapper.effect?.setFloat("layer", numLayers !== 1 ? layer / (numLayers - 1) : 0);
+                    this._drawWrapper.effect?.setInt("layerNum", layer);
+                    for (const name in this._textures) {
+                        this._drawWrapper.effect!.setTexture(name, this._textures[name]);
+                    }
                 }
 
-                // VBOs
-                engine.bindBuffers(this._vertexBuffers, this._indexBuffer, this._drawWrapper.effect!);
-
                 // Clear
                 if (this.autoClear) {
                     engine.clear(scene.clearColor, true, false, false);
                 }
 
                 // Draw order
                 engine.drawElementsType(Material.TriangleFillMode, 0, 6);
+                // Unbind and restore viewport
+                engine.unBindFramebuffer(this._rtWrapper, this.isCube || !this._generateMipMaps);

```

Comment:
Thank you.
Yes, the check for `_generateMipMaps` was a hack to prevent the mips from being automatically generated. I needed `generateMipMaps` to be set for the texture to behave as if it had mips but I create the mips myself and copy them in later so having them automatically generated by WebGL is a waste.
Maybe there's a better way of doing this?